Prepare KMS data encryption for migration to AES-GCM

enj opened the pull request:
This change updates the KMS envelope encryption to be able to use
AES-GCM to encrypt data using the DEK instead of AES-CBC.  To allow
for downgrades and HA upgrades, this functionality is only used on
reads.

In the next release we will be able to use AES-GCM for writes.
Backwards compatibility with previously encrypted resources will be
retained by falling back to AES-CBC on reads.

Signed-off-by: Monis Khan <mok@vmware.com>

/kind bug
xref #81127

/cc @liggitt @immutableT @smarterclayton 
@kubernetes/sig-auth-pr-reviews 
/priority important-longterm

```release-note
NONE
```

liggitt commented:
Doesn't this break skewed HA API servers and rollback scenarios?

liggitt commented:
if you want to do this config-free (which I like), you have to just change the read path in 1.18 (decrypt with CBC, fall back to GCM), and wait until 1.19 to change the write path (change encrypt to GCM, move decrypt to GCM first, then fallback to CBC)

enj commented:
> Doesn't this break skewed HA API servers and rollback scenarios?

🤦 yeah I was not thinking about the HA case.  Let me fix this up to use GCM as read-only in 1.18 and then swap the order in 1.19.

enj committed:
Prepare KMS data encryption for migration to AES-GCM

This change updates the KMS envelope encryption to be able to use
AES-GCM to encrypt data using the DEK instead of AES-CBC.  To allow
for downgrades and HA upgrades, this functionality is only used on
reads.

In the next release we will be able to use AES-GCM for writes.
Backwards compatibility with previously encrypted resources will be
retained by falling back to AES-CBC on reads.

Signed-off-by: Monis Khan <mok@vmware.com>

mikedanese commented on test/integration/master/transformation_testcase.go:
```diff
@@ -51,6 +51,12 @@ const (
+	// this looks exactly the same as the AES GCM secret but with a different value
+	legacyAESCBCKey       = "nbKcg2Dks/T/30ZCa29vLnyV/bhaFFnZ0cgVp7hagOE="
+	legacyEncryptedSecret = "azhzOmVuYzprbXM6djE6a21zLXByb3ZpZGVyOgAsbmJLY2cyRGtzL1QvMzBaQ2EyOXZMbnlWL2JoYUZGblowY2dWcDdoYWdPRT0RM8idSrtIpRoUN7uYzH+IjXOkMoCBO5NAtywkRIJG7PxQyVwO9jEA7vS6WEdJHbUIuElKqKMtb/EjZxxAwWZtJhZsr3kmUSAUxDDijCBHgKVvF/T9HzoOI4YXiFNPr04X5cQbyZIB2W4PG0lZAKVs9qqRlunObsUh9mbqGZYi+DAmMLISwldtHRVxktCD+QfuK6H1wS966Ts2r6TG42ELbuaoM9Dgg9JmU9UYc4+8CA=="
+	legacySecretVal       = "🤗🐼"
```
please nooooooo

mikedanese replied:
"\xf0\x9f\xa4\x97\xf0\x9f\x90\xbc"

enj replied:
😭

enj commented:
/milestone 1.18

enj commented:
/milestone v1.18

mikedanese commented:
This change makes sense to me. I'm not a fan of the current storage format of the encrypted storage format of these records. Given that we are doing a data migration here, I'm wondering if it makes sense to do more than just change the data encryption cipher.

immutableT commented:
@mikedanese Are you referring to something like [go-jose](https://github.com/square/go-jose) to produce an envelope in [JWE](https://tools.ietf.org/html/draft-ietf-jose-json-web-encryption-40#appendix-A) standard?
To do this, we need KMS Plugin to return metadata about how DEK was protected.

enj commented:
> This change makes sense to me. I'm not a fan of the current encrypted storage format of these records. Given that we are doing a data migration here, I'm wondering if it makes sense to do more than just change the data encryption cipher.

Could you elaborate on this some?  To frame the discussion, I have outlined what we store in etcd based on `master` today.

At ETCD path `/PATH_PREFIX/secrets/NAMESPACE/NAME` (called `HH` below)

The bytes stored are:

`k8s:enc:kms:v1:KMS_NAME:AABBCC`

`AA` = 2 byte (uint16 big-endian) length of encrypted DEK
`BB` = encrypted DEK (or reference to DEK)
`CC` = DEK(`GG`, `HH`) via AES256 CBC (this is unauthenticated so `HH` is ignored)

`CC` = `DDEEFF`

`DD` = IV
`EE` = Encrypted bytes
`FF` = Padding

`GG` = unencrypted bytes of API object
`HH` = the ETCD path, used as AEAD (but AES-CBC ignores this)

`PATH_PREFIX` configurable, it is normally `registry`.

mikedanese commented:
I think JWE, [CMS](https://tools.ietf.org/html/rfc5652#section-6), or a proto that we define would all be better than what we have now IMO. One thing that I would have liked at one point is some notion of key version within a single KMS plugin that could be encoded in the envelope.

Also I'm fishing for improvements right now since we have to make a format change anyway.

immutableT commented:
Using a standards-based envelope serialization (JWE or CMS) format has several benefits:
1. The envelope becomes self-describing, so during TransformFromStrorage we could apply the transform defined in the envelop
2. Users could easily recover their data from etcd (assuming they still have access to KMS) even if the cluster is not longer operational.
3. The logic of envelope construction will be outsourced to libraries like jose.
I think other scenarios (where data is encrypted outside the cluster) may become possible, but this requires more thought.

Both JWE and CMS expect the algorithm with which KEKs were encrypted. 
JWE example : {"alg":"RSA-OAEP","enc":"A256GCM"} 
KEK: RSA-RSA-OAEP
DEK: A256GCM

Today, KMS Plugin does not supply such information, so we would need to extend plugin's API. 
I don't think we need to wait for all kms-plugin authors to implement this change, instead we could fill "alg" with "unknown".

enj commented:
I did some digging to try and understand how we got to the current format.  We seem to have arrived at it somewhat organically - there were probably discussions in SIG Auth that were not "written down" (though I suppose a length prefixed encrypted DEK is about as simple as it gets).

https://github.com/kubernetes/community/pull/1581
https://github.com/kubernetes/enhancements/issues/460

https://github.com/kubernetes/kubernetes/pull/55684
https://github.com/kubernetes/kubernetes/issues/51965
https://github.com/kubernetes/kubernetes/pull/49350

https://docs.google.com/document/d/1S_Wgn-psI0Z7SYGvp-83ePte5oUNMr4244uanGLYUmw/

I am open to the idea for more radical changes here - lets discuss next year in the next SIG Auth meeting.

mikedanese commented:
If we did do something, it probably wouldn't be in this PR but it would make our lives easier if it were in the same release.

enj commented:
> If we did do something, it probably wouldn't be in this PR but it would make our lives easier if it were in the same release.

I am fine doing this as a step 0 (since it causes no behavioral change).  If we come up with something better before releasing 1.18, then we can just overwrite this code.

markyjackson-taulia commented:
Hi @enj 

Bug triage team here. Will this be making it in the v1.18 release? I see there are some conflicting files that need to be fixed.

markyjackson-taulia commented:
@enj Looks like we need a rebase

markyjackson-taulia commented:
@enj 
Gentle ping regarding rebase and see if this will still make it into the v1.18 release

enj commented:
/milestone clear

We need to work through some overarching issues in this area: https://github.com/kubernetes/kubernetes/pull/87438#issuecomment-584894679
